proxy/src/main/java/com/velocitypowered/proxy/command/VelocityCommandManager.java

```diff
@@ -69,107 +129,111 @@ public void unregister(final String alias) {
     if (command == null) {
       return false;
     }
 
+    I invocation = invocationFactory.createInvocation(command, source, alias, args);
     try {
-      if (!command.hasPermission(source, args)) {
+      if (!command.hasPermission(invocation)) {
         return false;
       }
-      command.execute(source, args);
+
+      command.execute(invocation);
       return true;
-    } catch (Exception e) {
+    } catch (final Exception e) {
+      if (e.getCause() instanceof CommandSyntaxException) {
+        // TODO Send invalid syntax message to player (exception contains details)
+        return false;
+      }
+
       throw new RuntimeException("Unable to invoke command " + cmdLine + " for " + source, e);
     }
   }
 
+  // Suggestions
 
-  @Override
-  public CompletableFuture<Boolean> executeAsync(CommandSource source, String cmdLine) {
-    CompletableFuture<Boolean> result = new CompletableFuture<>();
-    callCommandEvent(source, cmdLine).thenAccept(event -> {
-      CommandResult commandResult = event.getResult();
-      if (commandResult.isForwardToServer() || !commandResult.isAllowed()) {
-        result.complete(false);
-      }
-      String command = commandResult.getCommand().orElse(event.getCommand());
-      try {
-        result.complete(executeImmediately(source, command));
-      } catch (Exception e) {
-        result.completeExceptionally(e);
-      }
-    });
-    return result;
-  }
+  private <I extends CommandInvocation> CompletableFuture<List<String>> offerSuggestions(
+          final CommandSource source, final String alias, final String args) {
+    Command<I> command = getCommand(alias);
+    if (command == null) {
+      // No such command, so we can't offer any tab complete suggestions.
+      return CompletableFuture.completedFuture(ImmutableList.of());
+    }
 
-  @Override
-  public CompletableFuture<Boolean> executeImmediatelyAsync(CommandSource source, String cmdLine) {
-    Preconditions.checkNotNull(source, "source");
-    Preconditions.checkNotNull(cmdLine, "cmdLine");
-    CompletableFuture<Boolean> result = new CompletableFuture<>();
-    eventManager.getService().execute(() -> {
-      try {
-        result.complete(executeImmediately(source, cmdLine));
-      } catch (Exception e) {
-        result.completeExceptionally(e);
+    I invocation = invocationFactory.createInvocation(command, source, alias, args);
+    try {
+      if (!command.hasPermission(invocation)) {
+        return CompletableFuture.completedFuture(ImmutableList.of());
       }
-    });
-    return result;
-  }
 
-  public boolean hasCommand(String command) {
-    return commands.containsKey(command);
-  }
+      return command.suggestAsync(invocation).thenApply(ImmutableList::copyOf);
+    } catch (final Exception e) {
+      if (e.getCause() instanceof CommandSyntaxException) {
+        return CompletableFuture.completedFuture(ImmutableList.of());
+      }
 
-  public Set<String> getAllRegisteredCommands() {
```

Comment:
I would like to us to be able to inject proxy commands into tab-complete requests and into the Brigadier command tree if we can.

Reply:
I think this is already handled by https://github.com/VelocityPowered/Velocity/blob/21ad6a433f8b06d61de7e35bf1cf0658d5ca0f6e/proxy/src/main/java/com/velocitypowered/proxy/connection/client/ClientPlaySessionHandler.java#L400, `CommandManager#offerSuggestions` merges all suggestions from all command types, including `BrigadierCommand`s.

Reply:
But how about merging the proxy's Brigadier tree with that from the remote server?